### PR TITLE
adds tron support

### PR DESCRIPTION
**Pull request by MarcelBlockchain**
new PR to avoid months of rebasing.
Adds the missing `types`, `networkfees` and some small fixes.
Also adds data to `test/engine/fixtures.js`
I couldn't run the tests, got an`"Error: Cannot find module 'usb'"` from `ledgerhq` node_module. Anyone got this issue before ?

------
Notes to the updated PR:

The biggest change are the transaction fees. Every Tron user has a daily amount of free transactions. These are pulled using `trx_getAccountResource`. As long as the users `resources.freeNetLimit` is higher than `268`, the transaction is for free.

In case the account, the user sends tokens to is not yet activated, an activation fee of 0.1 TRX will be added. See more details below.

The costs for account activation and transaction per byte are pulled from the network using `async checkUpdateNetworkFees`. It’s unit is SUN (1 SUN = 0.000001 TRX) and are converted to TRX using `tronWeb.fromSun(myvalue)`. 

Acitivation fee for not activated receiving accounts: 100000 SUN

Transaction cost per byte: 10 SUN

For a tron transfer you need to pay for 268 bytes (`raw_data_hex.length`)

These base values won’t be changed anytime soon.

From the docs:

"The TRON network grants every account a free pool of Bandwidth Points for free transactions every 24 hours. To engage in transactions more frequently requires freezing TRX for additional bandwidth points, or paying the fee in TRX.

There are 5000 bandwidth points for free per account per day. When an account hasn’t frozen any balance, or when its bandwidth points have run out, complimentary bandwidth points can be used”

https://developers.tron.network/docs/bandwith

---

In case you want to start testing on shasta test net with free money, prepend 'shasta' to the link in tronEngine.js, tronPlugin.js and tronInfo.js

const tronWeb = new TronWeb({
  fullHost: 'https://api.trongrid.io'
})

--> api.shasta.trongrid.io

free shasta money: https://www.trongrid.io/shasta

Tron Metamask Alternative: https://github.com/TronLink/tronlink-extension

### CODE REVIEW

File: src/tron/tronEngine.js [hunk 1]
```diff
@@ -0,0 +1,641 @@
...
+      // if the receiver's account hasn't been activated yet
+      // pay the activiation fee
+      if (typeof jsonObjReceiver.freeNetLimit === 'undefined') {
+        totalFees = bns.add(
+          totalFees,
+          this.otherData.networkFees.createAccountFee
+        )
+      }
+    } catch (e) {
+      this.log('calcTxFee error: ', e)
```
Comment (peachbits) [inline src/tron/tronEngine.js:443]: Looks like this needs to throw?

File: src/tron/tronEngine.js [hunk 2]
```diff
@@ -0,0 +1,641 @@
...
+        const networkFees: TronNetworkFees = {
+          createAccountFee: tronWeb.fromSun(
+            jsonObj.find(e => e.key === 'getCreateAccountFee').value
+          ),
+          transactionFee: tronWeb.fromSun(
+            jsonObj.find(e => e.key === 'getTransactionFee').value
+          )
+        }
+        // only update network fee if it's not the same
+        if (this.otherData.networkFees !== networkFees) {
```
Comment (peachbits) [inline src/tron/tronEngine.js:377]: If will always be true. Either remove if statement or check actual values inside object

networkFees.createAccountFee !== this.otherDate.networkFees.createAccountFee
networkFees.transactionFee !== this.otherDate.networkFees.transactionFee

File: src/tron/tronEngine.js [hunk 3]
```diff
@@ -0,0 +1,641 @@
...
+      let otherParams: TronTxOtherParams = {
+        visible: false,
+        txID: '',
+        raw_data: {},
+        raw_data_hex: ''
+      }
+      // if TRON
+      if (currencyCode === PRIMARY_CURRENCY) {
+        // builds tx, but does NOT send it
+        const trxParams: TronTxOtherParams = await tronWeb.transactionBuilder.sendTrx(
```
Comment (peachbits) [inline src/tron/tronEngine.js:548]: It appears that the calculated networkFee isn't being included in the transactionBuilder. Where is the transactionBuilder getting the fee? There's a concern where the actual fee used will not match what we show the user.

**Comment by peachbits** [comment]
derivePublicKey test is failing

**Comment by MarcelBlockchain** [comment]
derivePublicKey test no more failing

**Comment by paullinator** [comment]
@MarcelBlockchain Thanks for the fixes. We're pushing this to QA testing now, and we'll merge once things look good.

**Comment by peachbits** [comment]
@MarcelBlockchain creating a new wallet results in a stuck wallet loading spinner at 50%. Could be related to a network call that's dependent on an address already being used and is failing therefore it isn't progressing the spinner.

**Comment by peachbits** [comment]
Completed in https://github.com/EdgeApp/edge-currency-accountbased/pull/464

**Pull request closed without merging**